Overhaul RUSTFLAGS handling for `compiler/` crates

nnethercote opened the pull request:
Currently some flags are put into RUSTFLAGS for all `compiler/` crates, but RUSTFLAGS is ignored for proc macro crates. 

Also, most compiler crates have `#![warn(unreachable_pub)]` in them, but a few are missing, and it's easy to overlook if a new crate is added.

This PR introduces a neat way to apply a flag to *all* `compiler/` crates, and uses it to replace all the `#![warn(unreachable_pub)]` attributes with a single `-Wunreachable-pub` mention.

r? @jieyouxu

nnethercote committed:
Add `-Wunreachable_pub` to `FLAGS_FOR_RUSTC`.

And fix the new errors in the handful of crates that didn't have a
`#![warn(unreachable_pub)]`.

nnethercote committed:
Pass certain flags when building all `compiler/` crates.

Currently they are passed to the ones satisfying `Mode::Rustc` via
RUSTFLAGS. But RUSTFLAGS is ignored for proc macro crates. This commit
makes those flags get added directly to the command for proc macro
crates.

nnethercote committed:
Remove `#![warn(unreachable_pub)]` from all `compiler/` crates.

It's no longer necessary now that `-Wunreachable_pub` is being passed.

nnethercote commented:
Best reviewed one commit at a time.

nnethercote commented:
#137930 is another PR where this new functionality will immediately be useful.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13669261975/job/38216276632) [(plain)](https://github.com/rust-lang/rust/commit/e14bcdc23276a0bc287318161787769b83b9302f/checks/38216276632/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking termcolor v1.4.1
    Checking home v0.5.9
    Checking build_helper v0.1.0 (/checkout/src/build_helper)
    Checking xz2 v0.1.7
error: constants have by default a `'static` lifetime
    |
    |
118 | pub const FLAGS_FOR_RUSTC: &[&'static str] = &[
    |                              -^^^^^^^---- help: consider removing `'static`: `&str`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_static_lifetimes
    = note: `-D clippy::redundant-static-lifetimes` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::redundant_static_lifetimes)]`

```

</details>

jieyouxu commented:
r? @onur-ozkan (for the rustc wrapper bit, I'd appreciate a second pair of eyes,  or reroll
r? jieyouxu

RalfJung commented on src/bootstrap/src/bin/rustc.rs:
```diff
@@ -145,6 +145,35 @@ fn main() {
 
+    // In `cargo.rs` we add the `FLAGS_FOR_RUSTC` flags for all `compiler/`
+    // crates to RUSTFLAGS. However, RUSTFLAGS is ignored for `compiler/` proc
+    // macro crates, because RUSTFLAGS is ignored when `--target` is given. So,
```
> RUSTFLAGS is ignored when `--target` is given

That's not quite true, it is not ignored. But it only applies to target crates, not to host crates.

RalfJung commented on src/bootstrap/src/bin/rustc.rs:
```diff
@@ -145,6 +145,35 @@ fn main() {
+    let crate_type = parse_value_from_args(&orig_args, "--crate-type");
+    if crate_type == Some("proc-macro") {
+        if let Ok(rustflags) = env::var("RUSTFLAGS") {
+            // Count how many of flags from `FLAGS_FOR_RUSTC` are in RUSTFLAGS.
```
Why do we have to use a fragile hacky heuristic like that? That shouldn't be needed. Bootstrap should clearly tell the wrapper what to add where, we should never do guesswork like this.

nnethercote replied:
I know very little about bootstrap and this was something I managed to get working. I'm happy to hear other ideas and implement them but I'll need a more detailed explanation than "bootstrap should clearly tell the wrapper" - how and where would this be done? :)

RalfJung replied:
I was imagining something like this, which already exists:

https://github.com/rust-lang/rust/blob/c85d3f6e874acd35647617161fd2ae950b94dedd/src/bootstrap/src/bin/rustc.rs#L138-L140

From some of the other discussion it sounds like what you dislike about this approach is that it will pass the flags to all invocations, not just the ones for workspace crates. However, non-workspace crates run with `--cap-lints` so that should be entirely non-consequential?

The discussion about this seems to be spread across multiple Zulip topics, issues, and PRs, so I have no idea which options have been considered. Has `RUSTC_LINT_FLAGS` been discarded for some reason?

nnethercote replied:
[This thread](https://rust-lang.zulipchat.com/#narrow/channel/233931-t-compiler.2Fmajor-changes/topic/Enable.20.60warn.28unused_crate_dependencies.29.20o.E2.80.A6.20compiler-team.23844) has most of the discussion.

I didn't even know RUSTC_LINT_FLAGS was a thing until just now. Like I said above (and in the thread), I know very little about bootstrap and I've tried several different approaches based on suggestions people have made at different times.

RalfJung replied:
That's fair -- I just thought bootstrap people were involved in the discussion / review so I'm a bit surprised it took so long for someone to suggest this. Let's make sure to include bootstrap people in the next PR.

RalfJung replied:
I guess I am indeed underestimating how fragmented bootstrap knowledge is.

Kobzol replied:
I also had no idea about the flag :) bootstrap's build logic is distributed amongst several places, it's full of hacks and edge cases, and hidden invariants that can break when you modify completely unrelated code.

RalfJung replied:
As I said, I didn't know about the flag either. I was about to suggest to add something like this flag, went looking to the sources for inspiration on how to call it, and then saw that it already existed.

jieyouxu replied:
Same 😆

cuviper commented on src/bootstrap/src/utils/shared_helpers.rs:
```diff
@@ -113,3 +113,14 @@ pub fn parse_value_from_args<'a>(args: &'a [OsString], key: &str) -> Option<&'a
+    "-Wkeyword_idents_2024",
+    "-Wunreachable_pub",
+    "-Wunsafe_op_in_unsafe_fn",
+];
```
It seems to me these should be in a `[workspace.lints]` table, although I think you still have to opt-in `[lints] workspace = true` in each crate.

nnethercote replied:
In every `compiler/` crate? Ugh, that's the kind of thing I'm trying to get rid of in this PR :(

cuviper replied:
It would be a one-time, one-line addition to each `Cargo.toml`, and it's a more principled way to control lints. IMO we should only use `RUSTFLAGS` (or the like) when Cargo can't do it at all.

It might even speed up the build to avoid the extra lints on out-of-tree dependencies! e.g. right now in a verbose build, I see "warning: prefer `FxHashMap` over `HashMap`" lints on a bunch of dependencies, where it's not actually applicable. Those are masked by cap-lints, but better not to run them at all.

nnethercote replied:
#138084 implements this. Unfortunately it doesn't work for rustc-specific lints like `rustc::internal`. Still, better than nothing, maybe.

jieyouxu commented:
Bootstrap thread: https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/Consistent.20handling.20of.20compiler.2F*.20RUSTFLAGs.20and.20lints

bors commented:
:umbrella: The latest upstream changes (presumably #138267) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
After the failure of the workspace lints approach (merged in #138084, reverted in #138304 for breaking miri and other external tools) I have implemented a `RUSTC_LINT_FLAGS`-based approach in #138331 which has the light footprint of this PR (i.e. doesn't touch many lines of code) while being much less hacky.
